File: python/elizaos_plugin_polymarket/actions/balances.py
"""
Balance actions for Polymarket.
"""

from __future__ import annotations

import importlib
from collections.abc import Callable
from typing import Protocol, cast

from elizaos_plugin_polymarket.error import PolymarketError, PolymarketErrorCode
from elizaos_plugin_polymarket.providers import get_clob_client
from elizaos_plugin_polymarket.types import BalanceAllowance


class RuntimeProtocol(Protocol):
    """Protocol for agent runtime."""

    def get_setting(self, key: str) -> str | None:
        """Get a setting value."""
        ...


def _coerce_str(value: object) -> str | None:
    if isinstance(value, str):
        return value
    if isinstance(value, int | float):
        return str(value)
    return None


def _resolve_asset_type(label: str) -> object:
    try:
        types_mod = importlib.import_module("py_clob_client.clob_types")
        asset_type = getattr(types_mod, "AssetType", None)
        if asset_type is not None:
            resolved = getattr(asset_type, label, None)
            if resolved is not None:
                return resolved
    except Exception:
        return label
    return label


def _get_balance_allowance(
    client: object,
    asset_type: object,
    token_id: str | None = None,
) -> BalanceAllowance:
    get_balance_allowance = getattr(client, "get_balance_allowance", None)
    if not callable(get_balance_allowance):
        get_balance_allowance = getattr(client, "getBalanceAllowance", None)
    if not callable(get_balance_allowance):
        raise PolymarketError(
            PolymarketErrorCode.API_ERROR,
            "get_balance_allowance method not available in CLOB client",
        )

    response_obj = cast(Callable[..., object], get_balance_allowance)(
        asset_type=asset_type, token_id=token_id
    )
    response = response_obj if isinstance(response_obj, dict) else {}
    balance = _coerce_str(response.get("balance")) or "0"
    allowance = _coerce_str(response.get("allowance")) or "0"
    return BalanceAllowance(balance=balance, allowance=allowance)


async def get_balances(
    token_ids: list[str] | None = None,
    include_collateral: bool = True,
    include_conditional: bool = True,
    max_tokens: int = 25,
    runtime: RuntimeProtocol | None = None,
) -> dict[str, object]:
    """
    Get collateral (USDC) and conditional token balances for a wallet.

    Args:
        token_ids: Optional list of token IDs for conditional balances
        include_collateral: Include collateral USDC balance
        include_conditional: Include conditional token balances
        max_tokens: Max number of token balances to fetch
        runtime: Optional agent runtime for settings

    Returns:
        Dictionary containing collateral balance and per-token balances

    Raises:
        PolymarketError: If fetching balances fails
    """
    if not include_collateral and not include_conditional:
        raise PolymarketError(
            PolymarketErrorCode.INVALID_ORDER,
            "At least one of include_collateral or include_conditional must be true",
        )

    try:
        client = get_clob_client(runtime)
        collateral: BalanceAllowance | None = None
        token_balances: dict[str, BalanceAllowance] = {}

        if include_collateral:
            asset_type = _resolve_asset_type("COLLATERAL")
            collateral = _get_balance_allowance(client, asset_type)

        if include_conditional:
            asset_type = _resolve_asset_type("CONDITIONAL")
            resolved_tokens = token_ids or []
            if max_tokens > 0:
                resolved_tokens = resolved_tokens[:max_tokens]
            for token_id in resolved_tokens:
                token_balances[token_id] = _get_balance_allowance(client, asset_type, token_id)

        return {
            "collateral": collateral,
            "token_balances": token_balances,
            "token_count": len(token_balances),
        }
    except PolymarketError:
        raise
    except Exception as e:
        raise PolymarketError(
            PolymarketErrorCode.API_ERROR,
            f"Failed to fetch balances: {e}",
            cause=e,
        ) from e

File: rust/src/actions/balances.rs
//! Dedicated balance action module for Polymarket.
//!
//! Provides wallet balance queries for both collateral (USDC) and
//! conditional tokens, with formatting helpers for display.

use crate::client::ClobClient;
use crate::error::{PolymarketError, Result};
use crate::types::BalanceAllowance;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// =============================================================================
// Types
// =============================================================================

/// Configuration for a balance query.
#[derive(Debug, Clone, Default)]
pub struct BalanceQueryParams {
    /// Specific token IDs to query conditional balances for.
    pub token_ids: Option<Vec<String>>,
    /// Whether to include collateral (USDC) balance.
    pub include_collateral: bool,
    /// Whether to include conditional token balances.
    pub include_conditional: bool,
    /// Maximum number of token balances to fetch.
    pub max_tokens: usize,
}

impl BalanceQueryParams {
    /// Create default params that include everything.
    pub fn all() -> Self {
        Self {
            token_ids: None,
            include_collateral: true,
            include_conditional: true,
            max_tokens: 25,
        }
    }

    /// Create params for collateral-only query.
    pub fn collateral_only() -> Self {
        Self {
            token_ids: None,
            include_collateral: true,
            include_conditional: false,
            max_tokens: 0,
        }
    }
}

/// Result of a balance query.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BalanceResult {
    /// Collateral (USDC) balance and allowance.
    pub collateral: Option<BalanceAllowance>,
    /// Conditional token balances keyed by token ID.
    pub token_balances: HashMap<String, BalanceAllowance>,
    /// Number of token balances returned.
    pub token_count: usize,
}

// =============================================================================
// Balance Query
// =============================================================================

/// Get wallet balances for collateral and/or conditional tokens.
///
/// # Arguments
///
/// * `client` - Authenticated CLOB client
/// * `params` - Balance query parameters
///
/// # Returns
///
/// Balance result containing collateral and token balances.
///
/// # Errors
///
/// Returns `PolymarketError` if:
/// - Client is not authenticated
/// - Neither collateral nor conditional is requested
/// - API request fails
pub async fn get_wallet_balances(
    client: &ClobClient,
    params: &BalanceQueryParams,
) -> Result<BalanceResult> {
    if !params.include_collateral && !params.include_conditional {
        return Err(PolymarketError::invalid_order(
            "At least one of include_collateral or include_conditional must be true",
        ));
    }

    if !client.has_credentials() {
        return Err(PolymarketError::auth_error(
            "API credentials required for balance queries",
        ));
    }

    let mut result = BalanceResult {
        collateral: None,
        token_balances: HashMap::new(),
        token_count: 0,
    };

    if params.include_collateral {
        let balance = client.get_collateral_balance().await?;
        result.collateral = Some(balance);
    }

    if params.include_conditional {
        if let Some(ref tokens) = params.token_ids {
            let limited = if params.max_tokens > 0 {
                &tokens[..tokens.len().min(params.max_tokens)]
            } else {
                tokens.as_slice()
            };

            for token_id in limited {
                match client.get_conditional_balance(token_id).await {
                    Ok(balance) => {
                        result.token_balances.insert(token_id.clone(), balance);
                    }
                    Err(e) => {
                        eprintln!(
                            "Warning: Failed to get balance for token {}: {}",
                            token_id, e
                        );
                    }
                }
            }
        }
        result.token_count = result.token_balances.len();
    }

    Ok(result)
}

/// Format balance result for display.
pub fn format_balance_result(result: &BalanceResult) -> String {
    let mut lines = vec!["💰 **Wallet Balances**\n".to_string()];

    if let Some(ref collateral) = result.collateral {
        lines.push(format!(
            "**Collateral (USDC):** {} (allowance: {})",
            collateral.balance, collateral.allowance
        ));
    }

    if !result.token_balances.is_empty() {
        lines.push(format!(
            "\n**Conditional Tokens:** ({} tokens)\n",
            result.token_count
        ));
        for (token_id, ba) in &result.token_balances {
            let short_id = if token_id.len() > 16 {
                format!("{}...", &token_id[..16])
            } else {
                token_id.clone()
            };
            lines.push(format!(
                "• `{}` — balance: {}, allowance: {}",
                short_id, ba.balance, ba.allowance
            ));
        }
    } else if result.token_count == 0 && result.collateral.is_some() {
        lines.push("\nNo conditional token balances found.".to_string());
    }

    lines.join("\n")
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_balance_query_params_all() {
        let params = BalanceQueryParams::all();
        assert!(params.include_collateral);
        assert!(params.include_conditional);
        assert_eq!(params.max_tokens, 25);
        assert!(params.token_ids.is_none());
    }

    #[test]
    fn test_balance_query_params_collateral_only() {
        let params = BalanceQueryParams::collateral_only();
        assert!(params.include_collateral);
        assert!(!params.include_conditional);
        assert_eq!(params.max_tokens, 0);
    }

    #[test]
    fn test_balance_result_serialization() {
        let result = BalanceResult {
            collateral: Some(BalanceAllowance {
                balance: "100.50".to_string(),
                allowance: "1000000".to_string(),
            }),
            token_balances: HashMap::from([(
                "token123".to_string(),
                BalanceAllowance {
                    balance: "50".to_string(),
                    allowance: "50".to_string(),
                },
            )]),
            token_count: 1,
        };

        let json = serde_json::to_string(&result).expect("serialization");
        assert!(json.contains("100.50"));
        assert!(json.contains("token123"));

        let deser: BalanceResult = serde_json::from_str(&json).expect("deserialization");
        assert_eq!(deser.collateral.unwrap().balance, "100.50");
        assert_eq!(deser.token_count, 1);
    }

    #[test]
    fn test_format_balance_result_with_collateral() {
        let result = BalanceResult {
            collateral: Some(BalanceAllowance {
                balance: "250.00".to_string(),
                allowance: "1000000".to_string(),
            }),
            token_balances: HashMap::new(),
            token_count: 0,
        };
        let formatted = format_balance_result(&result);
        assert!(formatted.contains("Wallet Balances"));
        assert!(formatted.contains("250.00"));
        assert!(formatted.contains("Collateral (USDC)"));
    }

    #[test]
    fn test_format_balance_result_with_tokens() {
        let result = BalanceResult {
            collateral: None,
            token_balances: HashMap::from([(
                "0xabcdef1234567890abcdef1234567890".to_string(),
                BalanceAllowance {
                    balance: "100".to_string(),
                    allowance: "100".to_string(),
                },
            )]),
            token_count: 1,
        };
        let formatted = format_balance_result(&result);
        assert!(formatted.contains("Conditional Tokens"));
        assert!(formatted.contains("1 tokens"));
        assert!(formatted.contains("0xabcdef12345678...")); // truncated
    }

    #[test]
    fn test_format_balance_result_empty() {
        let result = BalanceResult {
            collateral: Some(BalanceAllowance {
                balance: "0".to_string(),
                allowance: "0".to_string(),
            }),
            token_balances: HashMap::new(),
            token_count: 0,
        };
        let formatted = format_balance_result(&result);
        assert!(formatted.contains("No conditional token balances"));
    }
}
